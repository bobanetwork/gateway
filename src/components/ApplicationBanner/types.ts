export interface IAppAlert {
  type: 'success' | 'warning' | 'info'
  key: string
  message?: string
  canClose?: boolean
  isHidden?: boolean
  Component?: React.ElementType<{ className?: string }>
}
