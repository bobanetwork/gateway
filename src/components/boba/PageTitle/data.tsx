export const titleData = [
  {
    title: 'History',
    slug: 'Look back on past transactions',
    path: '/history',
  },
  {
    title: 'Earn',
    slug: 'Stake more, Earn more',
    path: '/earn',
  },
  {
    title: 'Stake',
    slug: 'Stake BOBA and earn rewards.',
    path: '/stake',
  },
  {
    title: 'DAO',
    slug: 'Participate in voting on proposals concerning the future of Boba Network',
    path: '/dao',
  },
  {
    title: 'Ecosystem',
    slug: 'Explore our decentralized apps on Boba Network',
    path: '/ecosystem',
  },
  {
    title: 'Trade',
    slug: 'You can acquire BOBA by trading on the listed exchanges.',
    path: '/trade',
  },
]