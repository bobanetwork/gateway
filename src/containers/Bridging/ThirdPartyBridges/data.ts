import BanxaLogo from 'assets/images/bridges/banxa.svg'
import BeamerLogo from 'assets/images/bridges/beamer.png'
import BoringDaoLogo from 'assets/images/bridges/Boringdao.png'
import ChainswapLogo from 'assets/images/bridges/chainswap.png'
import ConnextLogo from 'assets/images/bridges/connext.png'
import LayerswapLogo from 'assets/images/bridges/layerswap.png'
import MultichainLogo from 'assets/images/bridges/multichain.png'
import RangoExchangeLogo from 'assets/images/bridges/rango_exchange.png'
import SynapseLogo from 'assets/images/bridges/synapse.png'

import CelerLogoDark from 'assets/images/bridges/dark/dark_celer.png'
import RubicExchangeLogoDark from 'assets/images/bridges/dark/dark_rubic_exchange.png'
import ViaProtocolLogoDark from 'assets/images/bridges/dark/dark_via_protocol.png'
import CelerLogoLight from 'assets/images/bridges/light/light_celer.png'
import RubicExchangeLogoLight from 'assets/images/bridges/light/light_rubic_exchange.png'
import ViaProtocolLogoLight from 'assets/images/bridges/light/light_via_protocol.png'

export interface IBridges {
  name: string
  icon?: any
  iconLight?: any
  iconDark?: any
  link?: string
}

export const bobaBridges: IBridges[] = [
  {
    name: 'Banxa',
    icon: BanxaLogo,
    link: 'https://boba.banxa.com/',
  },
  {
    name: 'Beamer Bridge',
    icon: BeamerLogo,
    link: 'https://boba.network/project/beamer-bridge/',
  },
  {
    name: 'BoringDAO',
    icon: BoringDaoLogo,
    link: 'https://boba.network/project/boringdao/',
  },
  {
    name: 'Celer',
    icon: null,
    iconLight: CelerLogoLight,
    iconDark: CelerLogoDark,
    link: 'https://boba.network/project/celer/',
  },
  {
    name: 'Chainswap',
    icon: ChainswapLogo,
    link: 'https://boba.network/project/chainswap/',
  },
  {
    name: 'Connext',
    icon: ConnextLogo,
    link: 'https://boba.network/project/connext/',
  },
  {
    name: 'Layerswap',
    icon: LayerswapLogo,
    link: 'https://boba.network/project/layerswap-io/',
  },
  {
    name: 'Multichain',
    icon: MultichainLogo,
    link: 'https://boba.network/project/multichain/',
  },
  {
    name: 'Rango Exchange',
    icon: RangoExchangeLogo,
    link: 'https://boba.network/project/rango-exchange/',
  },
  {
    name: 'Rubic Exchange',
    icon: null,
    iconLight: RubicExchangeLogoLight,
    iconDark: RubicExchangeLogoDark,
    link: 'https://boba.network/project/rubic-exchange/',
  },
  {
    name: 'Synapse',
    icon: SynapseLogo,
    link: 'https://boba.network/project/synapse/',
  },
  {
    name: 'Via Protocol',
    icon: null,
    iconLight: ViaProtocolLogoLight,
    iconDark: ViaProtocolLogoDark,
    link: 'https://boba.network/project/via-protocol/',
  },
]
