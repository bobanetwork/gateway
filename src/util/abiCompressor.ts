/**
 * Script to prepare the human readable abi from the json.
 * based on the functions used out of the abi in network service.
 *
 * to prepare abi execute like
 * ```sh
 *  node ./util/abiCompressor.ts
 * ```
 *
 *
 */

import { ethers } from 'ethers'
import * as fs from 'fs'

const smart_contract_abi_paths = [
  './node_modules/@bobanetwork/core_contracts/artifacts/contracts/L2/messaging/L2StandardBridge.sol/L2StandardBridge.json',
  './node_modules/@bobanetwork/core_contracts/artifacts/contracts/standards/L2StandardERC20.sol/L2StandardERC20.json',
  './node_modules/@bobanetwork/core_contracts/artifacts/contracts/L2/predeploys/OVM_GasPriceOracle.sol/OVM_GasPriceOracle.json',
  './node_modules/@bobanetwork/contracts/artifacts/contracts/DiscretionaryExitFee.sol/DiscretionaryExitFee.json',
  './node_modules/@bobanetwork/contracts/artifacts/contracts/LP/L1LiquidityPool.sol/L1LiquidityPool.json',
  './node_modules/@bobanetwork/contracts/artifacts/contracts/Teleportation.sol/Teleportation.json',
  './node_modules/@bobanetwork/contracts/artifacts/contracts/LP/L2LiquidityPool.sol/L2LiquidityPool.json',
  './node_modules/@bobanetwork/contracts/artifacts/contracts/BobaFixedSavings.sol/BobaFixedSavings.json',
  './node_modules/@bobanetwork/contracts/artifacts/contracts/DAO/governance-token/BOBA.sol/BOBA.json',
  './node_modules/@bobanetwork/contracts/artifacts/contracts/DAO/governance/GovernorBravoDelegate.sol/GovernorBravoDelegate.json',
  './node_modules/@bobanetwork/contracts/artifacts/contracts/DAO/governance/GovernorBravoDelegator.sol/GovernorBravoDelegator.json',
  './node_modules/@bobanetwork/contracts/artifacts/contracts/L2BillingContract.sol/L2BillingContract.json',
  './src/deployment/contracts/L1ERC20.json',
  './src/deployment/contracts/NFTMonsterV2.json',
  './src/deployment/contracts/AuthenticatedFaucet.json',
  './src/deployment/contracts/WAGMIv0.json',
  './src/deployment/contracts/WAGMIv1.json',
  './src/deployment/contracts/crosschain/EthBridge.json',
  './src/deployment/contracts/crosschain/L2StandardERC20.json',
  './src/deployment/contracts/crosschain/LZEndpointMock.json',
]

const networkServicesStr = fs.readFileSync(
  './src/services/networkService.ts',
  'utf8'
)

const compress = () => {
  for (const path of smart_contract_abi_paths) {
    const rawData = fs.readFileSync(path, 'utf8')
    const abiObject = JSON.parse(rawData)
    const name = abiObject.contractName
    let abi = abiObject.abi
    abi = abi.filter((component) => {
      if (component['type'] === 'function' && component['name']) {
        if (networkServicesStr.includes(component['name'])) {
          return true
        }
        return false
      }
    })
    const abiInterface = new ethers.utils.Interface(abi)
    const humanReadableABI = abiInterface.format(ethers.utils.FormatTypes.full)
    const stringifiedHumanReadableABI = JSON.stringify(humanReadableABI)
    const result = `const ${name}ABI = ${stringifiedHumanReadableABI}\n export default ${name}ABI`
    fs.writeFile(`./src/services/abi/${name}.abi.ts`, result, () => {
      console.log(`wrote ${name}`)
    })
  }
}

;(() => {
  compress()
})()
