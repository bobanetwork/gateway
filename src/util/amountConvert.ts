/*
Copyright 2021-present Boba Network.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License. */

// we use BigNumber here for decimal support
import BigNumber from 'bignumber.js'
import { BigNumberish } from 'ethers'

export const logAmount = (
  amount: BigNumberish,
  power: number,
  truncate = 0
) => {
  // toString as some methods seem to supply BigNumber
  const x = new BigNumber(amount?.toString())
  const exp = new BigNumber(10).pow(power)

  const calculated = x.div(exp)

  if (truncate > 0) {
    return calculated.toFixed(truncate)
  } else {
    return calculated.toFixed()
  }
}

/*Takes a value such as 3.92 and converts it into
a BigNumber in wei

Duplicates

ethers.utils.parseUnits( valueString , decimalsOrUnitName )   =>   BigNumber
*/

export const powAmount = (amount: number, decimals: number) => {
  const x = new BigNumber(amount)
  const exp = new BigNumber(10).pow(decimals)

  const calculated = x.multipliedBy(exp)
  return calculated.toFixed(0)
}

/* more clearly named version of this */
export const toWei_String = (amount: number, decimals: number) => {
  const x = new BigNumber(amount)
  const exp = new BigNumber(10).pow(decimals)

  const calculated = x.multipliedBy(exp)
  return calculated.toFixed(0)
}

const lookupMap = {
  ETH: { provider: 'ethereum' },
  BOBA: { provider: 'boba-network' },
  OLO: { provider: 'oolongswap' },
  OMG: { provider: 'omisego' },
  USDC: { provider: 'usd-coin' },
  BNB: { provider: 'binancecoin' },
  tBNB: { provider: 'binancecoin' },
}

export const amountToUsd = (amount: number, lookupPrice, token) => {
  const { symbol } = token
  if (!symbol) {
    return 0
  }
  const currencyInfo = lookupMap[symbol]
  if (!!currencyInfo && !!lookupPrice[currencyInfo.provider]) {
    return amount * lookupPrice[currencyInfo.provider].usd
  } else if (!!lookupPrice[symbol.toLowerCase()]) {
    return amount * lookupPrice[symbol.toLowerCase()].usd
  } else {
    return 0
  }
}

export const formatLargeNumber = (num: number) => {
  const exp = Math.floor(Math.log10(Math.abs(num)) / 3)
  const scaledNum = Math.abs(num) / Math.pow(10, exp * 3)
  const formattedNumber = scaledNum.toFixed(2)
  const exponent = ['', 'k', 'M', 'B', 'T'][exp] || ''
  return isNaN(Number(formattedNumber)) ? '0' : formattedNumber + exponent
}
