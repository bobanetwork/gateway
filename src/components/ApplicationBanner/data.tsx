import React from 'react'
import { IAppAlert } from '.'
import { BannerText } from './styles'

export const bannerAlerts: IAppAlert[] = [
  {
    key: 'earn-banner-deprecation',
    type: 'warning',
    canClose: true,
    Component: () => (
      <BannerText>
        {' '}
        Boba Light bridge has been released and earn is sunset, to withdraw
        funds from liquidity <a href="/earn">CLICK HERE</a>{' '}
      </BannerText>
    ),
  },
]
