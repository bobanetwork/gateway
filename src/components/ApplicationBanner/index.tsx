import React, { Fragment, useEffect, useState } from 'react'
import {
  BannerWrapper,
  BannerAction,
  BannerContainer,
  BannerContent,
  BannerText,
  CloseIcon,
  CloseIconWrapper,
  BannerCenter,
  ArrowBtn,
} from './styles'
import { bannerAlerts } from './data'
import { IAppAlert } from './types'
import { useNetworkInfo } from 'hooks/useNetworkInfo'
import { ChevronLeft, ChevronRight } from '@mui/icons-material'

// @todo disable banner for sepolia.
// remove use of isAnchorageEnabled
const ApplicationBanner = () => {
  const [alerts, setAlerts] = useState<IAppAlert[]>([])
  const [storageChange, setStorageChange] = useState(false)
  const { isAnchorageEnabled } = useNetworkInfo()

  const [index, setIndex] = useState(0)

  useEffect(() => {
    const lastIndex = alerts.length - 1
    if (index < 0) {
      setIndex(lastIndex)
    }
    if (index > lastIndex) {
      setIndex(0)
    }
  }, [index, alerts])

  useEffect(() => {
    const sliderInterval = setInterval(() => {
      setIndex(index + 1)
    }, 7000)
    return () => clearInterval(sliderInterval)
  }, [index])

  useEffect(() => {
    const appBanners = bannerAlerts().map((alert) => {
      return {
        ...alert,
        isHidden: JSON.parse(
          localStorage.getItem(`appBanner__${alert.key}`) as string
        ),
      }
    })
    setAlerts(appBanners)
  }, [storageChange])

  const onClose = (alertKey: string) => {
    localStorage.setItem(`appBanner__${alertKey}`, JSON.stringify(true))
    setStorageChange(!storageChange)
  }

  if (!!isAnchorageEnabled) {
    return <></>
  }

  if (alerts && !alerts.length) {
    return <></>
  }

  return (
    <BannerWrapper>
      <BannerCenter>
        {alerts.map(
          (
            { key, canClose, type, message, Component, isHidden },
            alertIndex
          ) => {
            let position = 'nextSlide'
            if (alertIndex === index) {
              position = 'activeSlide'
            }
            if (
              alertIndex === index - 1 ||
              (index === 0 && alertIndex === alerts.length - 1)
            ) {
              position = 'lastSlide'
            }

            if (isHidden) {
              return <Fragment key={key}></Fragment>
            }
            return (
              <BannerContainer
                className={`${type} ${position}`}
                key={key}
                data-testid="banner-item"
              >
                <BannerContent>
                  {Component ? (
                    <Component />
                  ) : (
                    <BannerText>{message}</BannerText>
                  )}
                </BannerContent>
                {canClose && (
                  <BannerAction>
                    <CloseIconWrapper
                      data-testid={`close-icon-${key}`}
                      onClick={() => {
                        onClose(key)
                      }}
                    >
                      <CloseIcon />
                    </CloseIconWrapper>
                  </BannerAction>
                )}
              </BannerContainer>
            )
          }
        )}
        <ArrowBtn className="prev" onClick={() => setIndex(index - 1)}>
          <ChevronRight />
        </ArrowBtn>
        <ArrowBtn className="next" onClick={() => setIndex(index + 1)}>
          <ChevronLeft />
        </ArrowBtn>
      </BannerCenter>
    </BannerWrapper>
  )
}

export default ApplicationBanner
