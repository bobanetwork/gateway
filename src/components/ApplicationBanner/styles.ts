import { Svg, Typography } from 'components/global'
import styled from 'styled-components'
import Close from 'assets/images/close.svg'

export const BannerWrapper = styled('div')`
  display: flex;
  align-items: center;
  width: 100vw;
  max-width: 100%;
  flex-direction: column;
  justify-content: flex-start;
  gap: 2px;
`
export const BannerContainer = styled('div')`
  position: relative;
  transition: max-height 0.4s;
  padding: 5px 10px;
  overflow: hidden;
  display: flex;
  justify-content: center;
  align-items: center;
  gap: 10px;
  width: 100%;
  background: ${({ theme: { colors } }) => colors.green[300]};
  color: ${(props) => props.theme.primaryfg};

  &.open {
    max-height: 60px;
  }

  @media ${(props) => props.theme.screen.tablet} {
    padding: 0 10px;
    &.open.expand {
      max-height: 120px;
    }
  }
`
export const BannerContent = styled('div')``

export const BannerAction = styled('div')`
  cursor: pointer;
`

export const CloseIconWrapper = styled('div')`
  cursor: pointer;
  border-radius: 50%;
  &:hover {
    background: ${({ theme: { name, colors } }) =>
      name === 'light' ? colors.green[200] : colors.green[100]};
  }
  div {
    padding: 1px;
    height: 24px;
    width: 24px;
    display: flex;
    justify-content: center;
    align-items: center;
    border-radius: 50%;
  }
`
export const BannerText = styled(Typography).attrs({
  variant: 'body2',
})`
  text-align: center;
  max-width: 1440px;
  margin: 0 auto;
  padding: 10px 0;

  @media ${(props) => props.theme.screen.mobile} {
    padding: 5px 10px;
  }

  a {
    color: #000;
    text-decoration: underline;
    cursor: pointer;
    opacity: 0.65;
    text-transform: capitalize;
  }
`
export const CloseIcon = styled(Svg).attrs({
  src: Close,
  fill: '#fff',
})`
  height: 12px;
  width: auto;
  display: flex;
  align-items: center;
  justify-content: center;

  div {
    display: flex;
  }
  svg {
    max-width: 24px;
    min-width: 10px;
    height: auto;
    stroke: #000;
  }

  &:hover {
    cursor: pointer;
  }
`
