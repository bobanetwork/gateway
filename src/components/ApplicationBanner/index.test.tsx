import { fireEvent, render, screen } from '@testing-library/react'
import React from 'react'
import { Provider } from 'react-redux'
import { BrowserRouter } from 'react-router-dom'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import CustomThemeProvider from 'themes'
import { mockLocalStorage, mockedInitialState } from 'util/tests'
import ApplicationBanner from '.'
import { bannerAlerts } from './data'

jest.mock('./data', () => ({
  bannerAlerts: jest.fn(),
}))

const mockbannerAlerts = bannerAlerts as jest.MockedFunction<
  typeof bannerAlerts
>

const mockStore = configureStore([thunk])

Object.defineProperty(window, 'localStorage', {
  value: mockLocalStorage,
})

const renderApplicationBanner = () => {
  return render(
    <Provider store={mockStore(mockedInitialState)}>
      <BrowserRouter>
        <CustomThemeProvider>
          <ApplicationBanner />
        </CustomThemeProvider>
      </BrowserRouter>
    </Provider>
  )
}

describe('ApplicationBanner', () => {
  beforeEach(() => {
    // @ts-ignore
    mockbannerAlerts.mockImplementation(() => [
      {
        key: 'message-1',
        type: 'warning',
        canClose: true,
        Component: () => (
          <p>
            Message update 1 <a href="/update">CLICK HERE</a>
          </p>
        ),
      },
      {
        key: 'message-2',
        type: 'warning',
        canClose: true,
        message: 'message two goes here',
      },
    ])
  })

  test('should match snapshot when empty alerts', () => {
    mockbannerAlerts.mockImplementation(() => [])
    const { asFragment } = renderApplicationBanner()
    expect(asFragment()).toMatchSnapshot()
  })

  test('should match snapshot when alerts are enable', () => {
    const { asFragment } = renderApplicationBanner()
    expect(asFragment()).toMatchSnapshot()
  })

  test('should update localstorage and ui on clicking close', () => {
    renderApplicationBanner()
    const closeBtn = screen.getByTestId(`close-icon-message-1`)
    expect(closeBtn).toBeVisible()
    fireEvent.click(closeBtn)
    expect(localStorage.getItem(`appBanner__message-1`)).toEqual(
      JSON.stringify(true)
    )
    expect(screen.getAllByTestId('banner-item').length).toEqual(1)
  })
})
