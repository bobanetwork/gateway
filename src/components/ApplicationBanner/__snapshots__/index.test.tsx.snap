// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationBanner should match snapshot when alerts are enable 1`] = `
<DocumentFragment>
  <div
    class="sc-fPXMVe bmgsTt"
  >
    <div
      class="sc-gFqAkR hXrRXV warning"
      data-testid="banner-item"
    >
      <div
        class="sc-ikkxIA"
      >
        <p>
          Message update 1 
          <a
            href="/update"
          >
            CLICK HERE
          </a>
        </p>
      </div>
      <div
        class="sc-dAbbOL bqUnLe"
      >
        <div
          class="sc-feUZmu kDoDsn"
          data-testid="close-icon-message-1"
        >
          <div
            class="sc-hzhJZQ eNVlnR"
          >
            <div>
              <div
                data-src="close.svg"
                src=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-gFqAkR hXrRXV warning"
      data-testid="banner-item"
    >
      <div
        class="sc-ikkxIA"
      >
        <p
          class="sc-jXbUNg jCIOIT sc-fUnMCh iwiHIW"
        >
          message two goes here
        </p>
      </div>
      <div
        class="sc-dAbbOL bqUnLe"
      >
        <div
          class="sc-feUZmu kDoDsn"
          data-testid="close-icon-message-2"
        >
          <div
            class="sc-hzhJZQ eNVlnR"
          >
            <div>
              <div
                data-src="close.svg"
                src=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ApplicationBanner should match snapshot when empty alerts 1`] = `<DocumentFragment />`;
